### ISSUE TITLE
support `apt-get install jreleaser` as a way to install jreleaser

### ISSUE THREAD

**Issue by yogurtearl** [jreleaser/jreleaser#981]
**Is your feature request related to a problem? Please describe.**

For certain setups I use `apt-get` to install all packages. 

**Describe the solution you'd like**

would like `apt-get install jreleaser` to be a supported way of installing `jreleaser`

**Comment by aalmiray** [comment]
JReleaser currently produces `.rpm` & `.deb` installers for each release.
What would be needed to publish said installers to a location where `apt-get` could resolve them?
Bear in mind that these files contain binaries (no sources) thus the default location compatible with Ubuntu rules might not be possible.

**Comment by yogurtearl** [comment]
requesting ubuntu to add a new package is detailed here: https://wiki.ubuntu.com/UbuntuDevelopment/NewPackages

**Comment by ebourg** [comment]
You probably want to package it for Debian first, Ubuntu will synchronize the package into its repository automatically.

The first step is to check if the JReleaser dependencies are already packaged (the runtime dependencies and the non optional build plugins), and if some are missing, start by packaging them. Maven projects are fairly easy to package using the mh_make tool from the maven-debian-helper package.

Gradle projects are more difficult to package (especially if they use the Kotlin DSL, it isn't supported in Debian yet and the build has to be rewritten in Groovy), there is no tool to generate a package from a Gradle project. I recommend cloning an existing Gradle based package and use it as a template (for example a simple one like [opentest4j](https://salsa.debian.org/java-team/opentest4j)).

When the package is ready, you can contact the Debian Java Team (debian-java@lists.debian.org) and ask if someone can review and upload it. You can ping me if you need any help.

**Comment by aalmiray** [comment]
Thanks for the pointers @ebourg. As I understand Debian prefers source packages over binary packages, doesn't it? While it's possible to build JReleaser from source as per Debian's rules I think making it happen will take a while as not every dependency is already available in Debian. The build relies on Gradle, luckily it uses the Groovy DSL so that removes one stumbling block.

This being said, supporting `apt-get` is actually a two-fold goal:
1. Provide a `jreleaser` package
2. Support `apt` as package manager for any other project that relies on JReleaser for its releases.

Are there any alternatives for packaging binaries? Say like a community repository?

**Comment by ebourg** [comment]
> As I understand Debian prefers source packages over binary packages, doesn't it?

That's correct, providing source packages ensures that Debian users are able to modify the software they use, which is a core aspect of open source. And to achieve this goal a Debian distribution contains all the tools and dependencies required to build any of its components.

> While it's possible to build JReleaser from source as per Debian's rules I think making it happen will take a while as not every dependency is already available in Debian

If you have a list of the JReleaser dependencies I can check what is missing.

> Are there any alternatives for packaging binaries? Say like a community repository?

Ubuntu has the [PPA](https://help.launchpad.net/Packaging/PPA) (Personal Package Archives), there is no official equivalent for Debian. Package managers often support Debian packages (JFrog Artifactory and Sonatype Nexus do, Gitea too). But this always require editing the APT sources files (/etc/apt/sources.list), as well as trusting the repository owner not to publish packages that may disrupt the system.

If the project generates only one platform independent .deb file, providing a simple download link isn't that bad. An APT repository becomes interesting when the project generates many interdependent or platform specific .deb files.

**Comment by aalmiray** [comment]
> If you have a list of the JReleaser dependencies I can check what is missing.

At it's most basic, JReleaser's build requires Java 11 and Gradle 6.9.3, the latter bootstrapped via its wrapper (we don't store binaries in the repository). Every other JAR dependency is pulled from Maven Central. I'd assume this would be enough to build form source following Debina's requirements _unless_ JAR dependencies must also be found as Debian packages and not downloaded from Maven Central. FWIW these are the dependencies that JReleaser requires for running, however there are more required for building and testing

```
JavaEWAH-1.1.13.jar
annotations-2.20.16.jar
apache-client-2.20.16.jar
arns-2.20.16.jar
asn-one-0.6.0.jar
auth-2.20.16.jar
aws-core-2.20.16.jar
aws-query-protocol-2.20.16.jar
aws-xml-protocol-2.20.16.jar
bcpg-jdk15on-1.70.jar
bcpkix-jdk15on-1.70.jar
bcprov-jdk15on-1.70.jar
bcutil-jdk15on-1.70.jar
classmate-1.5.1.jar
commonmark-0.21.0.jar
commons-codec-1.15.jar
commons-compress-1.22.jar
commons-io-2.11.0.jar
commons-lang3-3.12.0.jar
commons-net-3.9.0.jar
commons-text-1.10.0.jar
compiler-0.9.10.jar
eddsa-0.3.0.jar
endpoints-spi-2.20.16.jar
eventstream-1.0.1.jar
failsafe-2.4.4.jar
feign-core-12.2.jar
feign-form-3.8.0.jar
feign-httpclient-12.2.jar
feign-jackson-12.2.jar
github-api-1.129.jar
http-client-spi-2.20.16.jar
httpclient-4.5.13.jar
httpcore-4.4.16.jar
jackson-annotations-2.14.2.jar
jackson-core-2.14.2.jar
jackson-databind-2.14.2.jar
jackson-dataformat-toml-2.14.2.jar
jackson-dataformat-xml-2.14.2.jar
jackson-dataformat-yaml-2.14.2.jar
jakarta.activation-2.0.1.jar
jakarta.mail-2.0.1.jar
jcl-over-slf4j-2.0.6.jar
json-utils-2.20.16.jar
jsonschema-generator-4.28.0.jar
jsonschema-module-jackson-4.28.0.jar
jzlib-1.1.3.jar
metrics-spi-2.20.16.jar
org.eclipse.jgit-5.13.0.202109080827-r.jar
picocli-4.7.1.jar
profiles-2.20.16.jar
protocol-core-2.20.16.jar
reactive-streams-1.0.3.jar
regions-2.20.16.jar
s3-2.20.16.jar
sdk-core-2.20.16.jar
slf4j-api-2.0.6.jar
slf4j-simple-2.0.6.jar
snakeyaml-1.33.jar
sshj-0.35.0.jar
stax2-api-4.2.1.jar
third-party-jackson-core-2.20.16.jar
tika-core-2.7.0.jar
twitter4j-core-4.0.7.jar
utils-2.20.16.jar
woodstox-core-6.5.0.jar
xz-1.9.jar
yamllint-1.5.0.jar
zstd-jni-1.5.4-1.jar
zt-exec-1.12.jar
```

> Ubuntu has the [PPA](https://help.launchpad.net/Packaging/PPA) (Personal Package Archives), there is no official equivalent for Debian. Package managers often support Debian packages (JFrog Artifactory and Sonatype Nexus do, Gitea too). But this always require editing the APT sources files (/etc/apt/sources.list), as well as trusting the repository owner not to publish packages that may disrupt the system.

Looks like Ubuntu PPAs do not allow pre-built binaries as per

> Note: We do not allow uploading pre-built binary packages. 

However I wonder if Ubuntu source builds follow the same rules as Debian or if they me be laxer that is, use the Gradle wrapper as provided by the build itself instead of using a pre-packaged Gradle installation provided by the system.

> If the project generates only one platform independent .deb file, providing a simple download link isn't that bad. An APT repository becomes interesting when the project generates many interdependent or platform specific .deb files.

We do provide `.deb` files which package ready-made binaries (https://github.com/jreleaser/jreleaser/releases/download/v1.5.0/jreleaser-installer_1.5.0-1_amd64.deb) which might be OK for Ubuntu but likely not for Debian.

**Comment by ebourg** [comment]
These dependencies are already packaged (the version may differ):

```
JavaEWAH-1.1.13.jar
bcpg-jdk15on-1.70.jar
bcpkix-jdk15on-1.70.jar
bcprov-jdk15on-1.70.jar
bcutil-jdk15on-1.70.jar
classmate-1.5.1.jar
commons-codec-1.15.jar
commons-compress-1.22.jar
commons-io-2.11.0.jar
commons-lang3-3.12.0.jar
commons-net-3.9.0.jar
commons-text-1.10.0.jar
compiler-0.9.10.jar (mustache?)
httpclient-4.5.13.jar
httpcore-4.4.16.jar
jackson-annotations-2.14.2.jar
jackson-core-2.14.2.jar
jackson-databind-2.14.2.jar
jackson-dataformat-toml-2.14.2.jar
jackson-dataformat-xml-2.14.2.jar
jackson-dataformat-yaml-2.14.2.jar
jakarta.activation-2.0.1.jar
jakarta.mail-2.0.1.jar
jcl-over-slf4j-2.0.6.jar
jzlib-1.1.3.jar
org.eclipse.jgit-5.13.0.202109080827-r.jar
picocli-4.7.1.jar
reactive-streams-1.0.3.jar
slf4j-api-2.0.6.jar
slf4j-simple-2.0.6.jar
snakeyaml-1.33.jar
stax2-api-4.2.1.jar
tika-core-2.7.0.jar
woodstox-core-6.5.0.jar
xz-1.9.jar
zstd-jni-1.5.4-1.jar
zt-exec-1.12.jar
```

And these ones are missing:

```
asn-one-0.6.0.jar
aws-sdk 2.20.16:
  annotations-2.20.16.jar
  apache-client-2.20.16.jar
  arns-2.20.16.jar
  auth-2.20.16.jar
  aws-core-2.20.16.jar
  aws-query-protocol-2.20.16.jar
  aws-xml-protocol-2.20.16.jar
  endpoints-spi-2.20.16.jar
  http-client-spi-2.20.16.jar
  json-utils-2.20.16.jar
  metrics-spi-2.20.16.jar
  profiles-2.20.16.jar
  protocol-core-2.20.16.jar
  regions-2.20.16.jar
  s3-2.20.16.jar
  sdk-core-2.20.16.jar
  third-party-jackson-core-2.20.16.jar
  utils-2.20.16.jar
commonmark-0.21.0.jar
eddsa-0.3.0.jar
eventstream-1.0.1.jar
failsafe-2.4.4.jar
feign-core-12.2.jar
feign-form-3.8.0.jar
feign-httpclient-12.2.jar
feign-jackson-12.2.jar
github-api-1.129.jar
jsonschema-generator-4.28.0.jar
jsonschema-module-jackson-4.28.0.jar
sshj-0.35.0.jar
twitter4j-core-4.0.7.jar
yamllint-1.5.0.jar
```

That's not insurmountable, but there is substantial work required.

> unless JAR dependencies must also be found as Debian packages and not downloaded from Maven Central

Indeed, no download is allowed at build time, the build jobs are run in offline mode, in containers with network disabled. That's also true for Ubuntu packages. Maven and Gradle are redirected to a system repository in `/usr/share/maven-repo` which is populated when the dependencies are installed.

> We do provide .deb files which package ready-made binaries which might be OK for Ubuntu but likely not for Debian.

I got a quick look at the file, you could remove the embedded JRE and depend on java11-runtime-headless instead. Debian and Ubuntu switched to OpenJDK 11 some time ago, and the next release will use OpenJDK 17 by default.

**Comment by aalmiray** [comment]
> That's not insurmountable, but there is substantial work required.

Indeed, as there are more dependencies required for compile & test but not runtime.

> Indeed, no download is allowed at build time, the build jobs are run in offline mode, in containers with network disabled. 

I was not aware of that, good to know, except of course the hard requirement for resolving all dependencies via the system repository.

> I got a quick look at the file, you could remove the embedded JRE and depend on java11-runtime-headless instead. Debian and Ubuntu switched to OpenJDK 11 some time ago, and the next release will use OpenJDK 17 by default.

This `.deb` file is created using `jpackage` (https://docs.oracle.com/en/java/javase/14/docs/specs/man/jpackage.html) which automatically embeds a copy of a Java Runtime to enable running on a matching platform without a previous installation of compatible Java runtime. Said JR may not necessarily be equivalent in size & capabilities as `jlink` (another JDK tool) is used to shrink the JR (by removing unneeded Java modules).

This being said, JReleaser could build yet another `.deb` file which has no bundled Java Runtime, but it would have to do so using a different mechanism (not via `jpackage`). This is certainly doable, just need to figure out the steps.

So, say such `.deb` file is produced and published. Given that it still provides a binary it wouldn't be acceptable to be distributed in a Debian repository, correct? But that would still make it easier for people to install it somehow, right? On that case, where should that `.deb` file be published to make it easier for consumers to find it? At the moment (as you've seen) it's linked as a release asset.

**Comment by ebourg** [comment]
> This being said, JReleaser could build yet another .deb file which has no bundled Java Runtime, but it would have to do so using a different mechanism (not via jpackage). This is certainly doable, just need to figure out the steps.

You could use [jdeb](https://github.com/tcurdt/jdeb) for that. Unless a project requires an extremely recent JDK (typically more recent than the latest LTS) it doesn't really make sense to embed the JRE into the .deb file.

> So, say such .deb file is produced and published. Given that it still provides a binary it wouldn't be acceptable to be distributed in a Debian repository, correct?

It wouldn't be acceptable for the official Debian repository, but it could be distributed through an altenative repository. That's what Jenkins does for example, they host their own APT repository https://pkg.jenkins.io/debian/

**Comment by aalmiray** [comment]
> You could use [jdeb](https://github.com/tcurdt/jdeb) for that. Unless a project requires an extremely recent JDK (typically more recent than the latest LTS) it doesn't really make sense to embed the JRE into the .deb file.

Exactly my thoughts, `jdeb` is a good candidate for this kind of thing. Sadly, there's no way to tell `jpackage` to skip bundling the JR as that is one of its main goals.

> It wouldn't be acceptable for the official Debian repository,

Understood. Only packages built from source will be accepted.

> but it could be distributed through an altenative repository. That's what Jenkins does for example, they host their own APT repository https://pkg.jenkins.io/debian/

Do you happen to know other APT repositories that may accept 3rd party packages? This might be a longshot.

**Comment by ebourg** [comment]
> Do you happen to know other APT repositories that may accept 3rd party packages?

There are several APT repositoty hosting services, some of them have a free tier. Gemfury for example seems to accept hosting public packages for free (I don't know if it's a time limited trial). I gave it a try for my pet project, [Jsign](https://github.com/ebourg/jsign), and it's really easy to use: Drag and drop the .deb file in the web interface, add `deb [trusted=yes] https://apt.fury.io/jsign/ /` in `/etc/apt/sources.list` to declare the repository, and just run `apt update && apt install jsign`. Done.

Another solution: since an APT repository is basically a set of passive HTTP resources, some people simply store the files in a dedicated Github Pages repository. See for example: https://assafmo.github.io/2019/05/02/ppa-repo-hosted-on-github.html. I'm not sure it complies with the Github ToS though.

**Comment by aalmiray** [comment]
Thank you for the pointers. Gemfury appears to offer a 14-day trial. It's FREE option may be for collaborators only, those that can join an existing account. We'll see how it goes with that service.

Alternatively I could host my own PPA at https://jreleaser.org which is **served** by Github Pages. Have to check the ToS.
